### PR TITLE
[exec] Improve execution logic, channel definitions and add functionality to substitute, transform and format Item values

### PR DESCRIPTION
**Pull request by kgoderis**
… substitute, transform and format Item values. This PR replaces the previous PR for the Exec 2.0 binding, and takes into account the feedback voiced by early adaptors. Applied the static code analysis tool, all proposed changes made except for header check

Signed-off-by: Karel Goderis <karel.goderis@me.com>

### CODE REVIEW

**Comment by kgoderis** [comment]
@steve-bate Hey Steve, you might want to have a look at this new PR. I have implemented your suggestion with respect to the Item value substitution, and clarified some other areas. Best is to start with the README.md as some things have changed

File: addons/binding/org.openhab.binding.exec/README.md [hunk 1]
```diff
@@ -4,54 +4,90 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
+     - the name of any Item
+     - `exec-input`, denoting the current State of the input Channel of the `command` Thing
+     - `exec-time`, denoting the current date (as java.util.Date)     
+- <transform> is any valid Transformation service expression, e.g. REGEX((.*?))
+- <formatter> is a formatting string using the well known syntax of the `java.util.Formatter` class
+
+At runtime the binding will grab the value of the <key>, transform it with the <transform> expression, and finally format the transformation result using the <formatter> format. For example,
+
+```
+Thing exec:command:lightcontroller [command="/usr/local/bin/light.sh  ${lightSwitch:MAP(en.map):%1$s}"]
```
Comment (steve-bate) [inline addons/binding/org.openhab.binding.exec/README.md]: Can the transformations return more than one value? Can the '1$' be removed?

File: addons/binding/org.openhab.binding.exec/README.md [hunk 2]
```diff
@@ -4,54 +4,90 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
+```
+
+When the [command] of the lightcontroller item is executed then the value of the lightSwitch Item is retrieved, transformed using the en.map MAP Transform, and passed on as a literal string (i.e. %1$s takes the first argument of the result of the Transform, and formats it as a String) to the ligh.sh script
```
Comment (steve-bate) [inline addons/binding/org.openhab.binding.exec/README.md]: There's a typographical error ('ligh.sh').

File: addons/binding/org.openhab.binding.exec/README.md [hunk 3]
```diff
@@ -4,54 +4,90 @@ This binding integrates the possibility to execute arbitrary shell commands.
...
+
+For each [command] a separate Thing has to be defined.
+
+```
+Thing exec:command:apc [command="/usr/local/bin/apcaccess  status", interval=15, timeout=5, runOnInput=true, repeatEnabled=true]
+```
+
+[command] itself can be enhanced using some specific qualifiers that will be substituted at runtime by actual values. The qualifiers have the following syntax:
+
+`${<key>:<transform>:<formatter>}`
```
Comment (steve-bate) [inline addons/binding/org.openhab.binding.exec/README.md]: Very nice!

File: addons/binding/org.openhab.binding.exec/README.md (hunk 1 above)
Comment (kgoderis) [inline addons/binding/org.openhab.binding.exec/README.md]: Yes, they can return whatever they want, and thus it can be 1$ 2$ ...

**Comment by kgoderis** [comment]
@steve-bate Hey Steve, did you already test the above code? would like to push it a bit forward

**Comment by steve-bate** [comment]
@kgoderis I probably won't have time to try it before the weekend. If you made the updated bundle jar available, I'm wondering if someone in the forums would be willing to try it sooner.

**Comment by steve-bate** [comment]
This is definitely an improvement. However, it seems like the motivating switch scenario, where the switch can be manually controlled, isn't working. My monitoring thing:

```
exec:command:switch_monitor [command="switch_control.sh check", interval=5] {
    Channels:
        Switch : output
}
```

The switch item can be successfully bound to the channel. However, the ExecHandler attempts to publish a string to the channel's items and a string is not acceptable data type for a SwitchItem. The logged information is:

```
2017-04-14 07:50:39.858 [DEBUG] [s.c.internal.items.ItemUpdater:72   ] - Received update of a not accepted type (StringType) for item ExecSwitch
```

The way I test this is that my `switch_control.sh check` script invocation returns a value from a file when it is invoked. That way I can externally set the value in the file to ON or OFF and see if the UI updates correctly. If it's helpful, here's my bash test script that can simulate an externally controlled switch:

```bash
#!/bin/bash

function log {
	echo $(date '+%H:%M:%S') $* >> /tmp/switch_log.txt
}

if [ "$1" == "check" ]
then
    if [ -e /tmp/switch_state.txt ]
    then
    	status=$(cat /tmp/switch_state.txt)
    	log "Check result: $status"
	else
		log "Check result (default): OFF"
		echo 'OFF'
	fi
else
	log "Setting state: $1"
	echo $1 > /tmp/switch_state.txt
fi
```

To simulate external switch control, I execute an `echo OFF > /tmp/switch_state.txt` (for example). The next invocation of the monitoring thing will return the new state.

**Comment by steve-bate** [comment]
As an experiment I added a Convertible interface to the StringType and added code to support converting it to an OnOffType. The switch example works but I don't know if there's a better way. @kaikreuzer, is there any existing support for doing automatic StringType conversions to other types, such as OnOffType, DecimalType and so on (other than adding Convertible support for all the types)?

**Comment by kaikreuzer** [comment]
@steve-bate No, and I do not really see a requirement for this...
What we have is the `getStateAs` method on the items - so if you have an item and want to have its state not as the primary type, but as something else, this should be used.
Additionally, there is also the `TypeParser`, which can be used to get a typed instance from a plain string (which you might have received from some external source).

**Comment by steve-bate** [comment]
> No, and I do not really see a requirement for this...
> ... there is also the TypeParser, which can be used to get a typed instance from a plain string (which you might have received from some external source)

@kaikreuzer Your response is bit perplexing, but with some extra code the TypeParser appears to satisfy the requirements of the ExecHandler (see the extensive discussions over the last several months for details about the requirements and issues related to to this binding). 

@kgoderis I was able to make the conversion work with the following code. This may not be the best way, but I'm primarily an OH user rather than a core developer so I'm not an expert on the OH internals.
```java
ChannelUID channelUID = new ChannelUID(getThing().getUID(), OUTPUT);
Set<Item> links = linkRegistry.getLinkedItems(channelUID);
if (links.size() > 0) {
   // All linked items must be the same type so use the first one for the output conversion.
   Item item = links.iterator().next();
   State outputState = TypeParser.parseState(item.getAcceptedDataTypes(), transformedResponse);
   updateState(channelUID, outputState);
}
```
My time available is limited, so I won't be able to contribute further feedback on this PR in the near future. I still recommend getting user feedback from the community forums.

**Comment by kgoderis** [comment]
@steve-bate i think Kai is rather referring to the usage of getStateAs in Rules than in the code of the binding. That is only a small work around and can be tailored by the end user. Nevertheless I will revisit the code asap and then I think we should put in the repo for wider adaption and feedback.

**Comment by martinvw** [comment]
@kgoderis is this ready to review, I saw some discussion going on. Should it either be WIP or can you resolve the merge conflicts than I ll queue it up for a review. Thanks for the contribution!

**Comment by kgoderis** [comment]
Eh? Rebased in a new branch, and forced pushed it to the old branch, and GitHub thinks it needs to close the PR

**Comment by kgoderis** [comment]
@martinvw I am sure you will find quite some changes to because the code dates from before the new coding guidelines and static code analysis tool

**Comment by kgoderis** [comment]
@martinvw I am unable to fix this PR. As it is, I get a conflict in 1 file and I do not understand why, as this involves a single line of code that is added. if I rebase the whole PR on the HEAD of the Master, the n I get an error on pom.xml which I do not understand either. Can you fix this PR for me?

**Comment by martinvw** [comment]
I rebased it, please make sure you do not merge it with your own version locally, either discard your local branch or use the following commands.

```
git fetch --all
git reset --hard origin/exec-enhance-bis
```

And some first findings since I wanted to compile it:

image

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 4]
```diff
@@ -53,191 +58,326 @@
...
+                        proc.destroyForcibly();
+                    }
 
-                updateState(RUN, OnOffType.OFF);
-                updateState(EXIT, new DecimalType(proc.exitValue()));
+                    updateState(RUN, OnOffType.OFF);
+                    updateState(EXIT, new DecimalType(proc.exitValue()));
 
-                outputBuilder.append(errorBuilder.toString());
+                    outputBuilder.append(errorBuilder.toString());
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Maybe just add them to same `StringBuilder` right away.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 5]
```diff
@@ -53,191 +58,326 @@
...
 
-                outputBuilder.append(errorBuilder.toString());
+                    String transformedResponse = StringUtils.chomp(outputBuilder.toString());
+                    String transformation = (String) getConfig().get(TRANSFORM);
 
-                String transformedResponse = StringUtils.chomp(outputBuilder.toString());
-                String transformation = (String) getConfig().get(TRANSFORM);
+                    if (transformation != null && transformation.length() > 0) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: StringUtils?

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 6]
```diff
@@ -53,191 +58,326 @@
...
-                updateState(OUTPUT, new StringType(transformedResponse));
+                    DateTimeType stampType = new DateTimeType(Calendar.getInstance());
+                    updateState(LAST_EXECUTION, stampType);
 
-                DateTimeType stampType = new DateTimeType(Calendar.getInstance());
-                updateState(LAST_EXECUTION, stampType);
+                }
 
+            } catch (Exception e) {
+                logger.error("An exception occurred while executing a command : '{}'", e.getMessage(), e);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Should you mention which command? Should you catch all exceptions, what kind do you expect? Should it be error level?

Note that the finally will even be triggered if we do not catch the exception.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 7]
```diff
@@ -53,191 +58,326 @@
...
+            } finally {
+                lock.unlock();
             }
         }
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Can this method 130+ lines be split up?

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 8]
```diff
@@ -53,191 +58,326 @@
...
+                String format = null;
+                String subkey = "";
+                if (key.contains(":")) {
+                    String parts[] = key.split(":");
+                    subkey = parts[0];
+
+                    if (parts.length == 2) {
+                        format = parts[1];
+                    } else if (parts.length == 3) {
+                        transform = parts[1];
```
[the remark this answers is not in the document]
Comment (kgoderis) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: @martinvw The formatting is done after the transformation, therefore I choose to put it last so that the user would read it that way, e.g. from left to right

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 9]
```diff
@@ -53,191 +58,326 @@
...
 
-                logger.trace("The command to be executed will be '{}'", commandLine);
+                if (commandLine != null && !commandLine.isEmpty()) {
+
+                    updateState(RUN, OnOffType.ON);
+
+                    // For some obscure reason, when using Apache Common Exec, or using a straight implementation of
+                    // Runtime.Exec(), on Mac OS X (Yosemite and El Capitan), there seems to be a lock race condition
+                    // randomly appearing (on UNIXProcess) *when* one tries to gobble up the stdout and sterr output of
+                    // the
```
[the remark this answers is not in the document]
Comment (kgoderis) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: The file is auto-formatted like this when saving it in Eclipse.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinvw) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Okay, if the error message and documentation are clear its fine with me

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (martinvw) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Yes, but it will not glue back single words, so maybe make it one line and let eclipse reformat?

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java (hunk 4 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: I prefer to keep them apart

File: addons/binding/org.openhab.binding.exec/README.md [hunk 10]
```diff
@@ -10,147 +10,93 @@ Currently, the binding supports a single type of Thing, being the `command` Thin
 
 The binding does not require any specific configuration.
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/README.md]: Please note that careful though went into rewriting parts of the README by the community, also a lot of small markdown errors where fixed looking at changes it seems you might have overwritten some of these changes.

Maybe @Josar can help you out here?

File: addons/binding/org.openhab.binding.exec/META-INF/MANIFEST.MF [hunk 11]
```diff
@@ -27,3 +29,4 @@ Import-Package:
  org.osgi.framework,
  org.slf4j
 Service-Component: OSGI-INF/*.xml
+Require-Bundle: org.eclipse.smarthome.core
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/META-INF/MANIFEST.MF]: The manifest must not contain any `Require-Bundle` entries. Instead, `Import-Package` must be used. [[B.4](http://docs.openhab.org/developers/development/guidelines.html#b-osgi-bundles)]

This means you also have to reference specific packages in the `Import-Package` instead of the top-level packages needed for `Require-Bundle`

File: addons/binding/org.openhab.binding.exec/README.md [hunk 12]
```diff
@@ -10,147 +10,93 @@ Currently, the binding supports a single type of Thing, being the `command` Thin
...
 
+Note that the commands are executed in the context and with the privileges of the process running the java virtual machine. It is not advised to run the virtual machine as superuser/root. The execution of the commands is triggered by either sending a Command or State Update to the input Channel of the `command` Thing, if configured so, or by sending the ON Command to the run Channel of the `command` Thing (see below)
```
Comment (Josar) [inline addons/binding/org.openhab.binding.exec/README.md]: We put this in the -**Linux:** section and added some more info. 
New is "The execution ..." which i think is not apropriately placed here, but a good additional information.
I would rephrase it and place it to .. maybe the channels? Under the tabular.
" The execution of the command can be triggerd by either sending the ON command to the run channel or by sending a command or state update to the input channel, when the thing is configured with runOnInput=true."

Comment (Josar) [reply addons/binding/org.openhab.binding.exec/README.md]: I will stop here as this will just tripple the work. my suggestion to @kgoderis:
Try to first merge the most obvious things and then i am happy to review it again.

File: addons/binding/org.openhab.binding.exec/README.md (hunk 10 above)
Comment (kgoderis) [reply addons/binding/org.openhab.binding.exec/README.md]: @martinvw @Josar I was just force pushing some mods based on the previous feedback round, and I need to further test things before it is completed. I also saw the Linux related changes in the main branch wrt readme.md, so I need to pick those up and re-integrate these in the doc

**Comment by kgoderis** [comment]
@martinvw the static code tool keeps crashing on the README.md, so apart from any changes proposed by the tool (possibly in addition to your screenshot here above), all other changes are now included in the PR.

**Comment by kaikreuzer** [comment]
@svilenvul Are you aware of this but in SAT and is it possibly already fixed in 0.4.0?

File: addons/binding/org.openhab.binding.exec/README.md (hunk 10 above)
Comment (martinvw) [reply addons/binding/org.openhab.binding.exec/README.md]: @kgoderis what is the status of your note above:

> @martinvw @Josar I was just force pushing some mods based on the previous feedback round, and I need to further test things before it is completed. I also saw the Linux related changes in the main branch wrt readme.md, so I need to pick those up and re-integrate these in the doc as I do not want to loose the community's inputs

**Comment by martinvw** [comment]
The markdown crash should be fixed now btw @kgoderis, if we can get rid of the last remaining points and the merge conflicts it might be time to wrap this one up as well :-)

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 13]
```diff
@@ -53,191 +58,305 @@
...
 
-                if (transformation != null && transformation.length() > 0) {
-                    transformedResponse = transformResponse(transformedResponse, transformation);
                 }
 
-                updateState(OUTPUT, new StringType(transformedResponse));
-
-                DateTimeType stampType = new DateTimeType(Calendar.getInstance());
-                updateState(LAST_EXECUTION, stampType);
+            } catch (
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: This formatting looks really bizar can you maybe reformat all the code for this binding using the default eclipse setup?

**Comment by martinvw** [comment]
@kgoderis a kind reminder if your force push, we will not be notified :-)

I just checked and saw some travis errors, if you agree that this done let's make sure its get outs there before the 2.3, hower I think this might contain some breaking changes, given we have a new way of informing please add to to both [the wiki](https://github.com/openhab/openhab-distro/wiki/%5BDRAFT%5D-openHAB-2.3-Release-Notes) and the [update warnings](https://github.com/openhab/openhab-distro/blob/master/distributions/openhab/src/main/resources/bin/update.lst)

```
[INFO] --- sat-plugin:0.4.1:report (default) @ org.openhab.binding.exec ---
[INFO] Individual report appended to summary report.
[ERROR] Code Analysis Tool has found: 
 2 error(s)! 
 4 warning(s) 
 4 info(s)
[WARNING] .binding.exec/ESH-INF/thing/thing-types.xml:[3]
There were whitespace characters used for indentation. Please use tab characters instead
[WARNING] .binding.exec/ESH-INF/binding/binding.xml:[3]
There were whitespace characters used for indentation. Please use tab characters instead
[WARNING] .binding.exec/pom.xml:[3]
There were whitespace characters used for indentation. Please use tab characters instead
[WARNING] .binding.exec/OSGI-INF/ExecHandlerFactory.xml:[3]
There were whitespace characters used for indentation. Please use tab characters instead
[ERROR] org.openhab.binding.exec.internal.ExecCommandConfiguration.java:[1]
Header is missing
[ERROR] org.openhab.binding.exec.internal.ExecCommandConfiguration.java:[5]
An author tag is missing
[INFO] Detailed report can be found at: file:////home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.exec/target/code-analysis/report.html
```

**Comment by martinvw** [comment]
@kgoderis are you able to pick up the SAT reportings, it would be great to get this wrapped up and included in the 2.3 version. Thanks!

**Comment by martinvw** [comment]
@kgoderis it would still be great to get this wrapped up, then it can get in the next release,

**Comment by wborn** [comment]
I've fixed the merge conflicts and build errors on this PR @martinvw, @kgoderis by rebasing and an extra commit.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 14]
```diff
@@ -53,191 +58,299 @@
...
+                        if (format != null) {
+                            return String.format(format, transformedInput);
+                        } else {
+                            return execInput;
+                        }
+                    } else {
+                        return null;
+                    }
+                } else {
+                    String transformedInput = itemRegistry.getItem(subkey).getState().toString();
```
Comment (wborn) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: Do you think the Exec binding should do these transformations itself using item state from the item registry @kaikreuzer?

**Comment by martinvw** [comment]
> I've fixed the merge conflicts and build errors on this PR @martinvw, @kgoderis by rebasing and an extra commit.

I remember that these changes are quite breaking, do they have the right balance of breaking added value at the moment?

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java (hunk 14 above)
Comment (martinvw) [reply addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: > Do you think the Exec binding should do these transformations itself using item state from the item registry @kaikreuzer?

@kaikreuzer any idea's? Or shall we first get rid of #1637

**Comment by wborn** [comment]
Here's a [org.openhab.binding.exec-2.4.0-SNAPSHOT.jar](https://github.com/wborn/openhab2-addons/releases/download/exec-enhance-bis/org.openhab.binding.exec-2.4.0-SNAPSHOT.jar) containing the changes from this PR @claell.

**Comment by claell** [comment]
@martinvw I think I did not have many problems with this binding or created issues about them. Am I missing something here (maybe #1637?) or did you mean someone else?

**Comment by martinvw** [comment]
No this is the big refactoring which stopped the PR you mentioned from being merged. It would be great if you could test this version

**Comment by claell** [comment]
Ah, did not know the merge of #1637 was blocked by something.

Unfortunately I am lacking time, setup and knowledge for an overall test of this version. What I can offer is a short run on Ubuntu to see, whether it works, but that will probably not help much, correct?

**Comment by martinvw** [comment]
That is also appreciated 👍

**Comment by claell** [comment]
Alright, I did a short test with the latest stable openHAB runtime and OpenJDK.

Configuration completely via PaperUI.

Repeated command execution works for the two I tested using an intervall.

Below is a screenshot of the PaperUI dashboard showing the items linked to the thing channels. I used the command "ls" to also test the String item with the response.
bildschirmfoto von 2018-11-17 14-37-08

One possible problem I experienced is the Switch item, which cannot manually be changed from PaperUI. Is this just thought as a feedback switch that can be used in rules?

If yes I do not see an easy way to manually trigger the execution from the Dashboard.

**Comment by martinvw** [comment]
> One possible problem I experienced is the Switch item, which cannot manually be changed from PaperUI. Is this just thought as a feedback switch that can be used in rules?
> 
> If yes I do not see an easy way to manually trigger the execution from the Dashboard.

@kgoderis can you answer this?

**Comment by Hilbrand** [comment]
@kgoderis Do you have any time, plans to continue these changes to the exec binding in this pr? I think they're a valuable additions and if you don't have any time/plans can I pick up and finish your work?

**Comment by J-N-K** [comment]
Cosing due to inactivity for more than six month.
